sdks/java/extensions/ml/src/main/java/org/apache/beam/sdk/extensions/ml/RecommendationAIImportCatalogItems.java

```diff
@@ -0,0 +1,166 @@
+import com.google.cloud.recommendationengine.v1beta1.CatalogName;
+import com.google.cloud.recommendationengine.v1beta1.CatalogServiceClient;
+import com.google.cloud.recommendationengine.v1beta1.ImportCatalogItemsRequest;
+import com.google.cloud.recommendationengine.v1beta1.ImportCatalogItemsResponse;
+import com.google.cloud.recommendationengine.v1beta1.InputConfig;
+import com.google.protobuf.util.JsonFormat;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.concurrent.ExecutionException;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PCollectionTuple;
+import org.apache.beam.sdk.values.TupleTag;
+import org.apache.beam.sdk.values.TupleTagList;
+import org.checkerframework.checker.nullness.qual.Nullable;
+import org.json.JSONObject;
+
+/**
+ * A {@link PTransform} connecting to the Recommendations AI API
+ * (https://cloud.google.com/recommendations) and creating {@link CatalogItem}s. *
+ *
+ * <p>Batch size defines how many items are created at once per batch (max: 5000).
+ *
+ * <p>The transform consumes {@link KV} of {@link String} and {@link GenericJson}s (assumed to be
+ * the catalog item id as key and contents as value) and outputs a PCollectionTuple which will
+ * contain the successfully created and failed catalog items.
+ *
+ * <p>It is possible to provide a catalog name to which you want to add the catalog item (defaults
+ * to "default_catalog").
+ */
+@AutoValue
+@SuppressWarnings({"nullness"})
+public abstract class RecommendationAIImportCatalogItems
+    extends PTransform<PCollection<KV<String, GenericJson>>, PCollectionTuple> {
+
+  public static final TupleTag<CatalogItem> SUCCESS_TAG = new TupleTag<CatalogItem>() {};
+  public static final TupleTag<CatalogItem> FAILURE_TAG = new TupleTag<CatalogItem>() {};
+
+  public static Builder newBuilder() {
+    return new AutoValue_RecommendationAIImportCatalogItems.Builder();
+  }
+
+  /** @return ID of Google Cloud project to be used for creating catalog items. */
+  public abstract String projectId();
+
+  /** @return Name of the catalog where the catalog items will be created. */
+  public abstract @Nullable String catalogName();
+
+  /** @return Size of input elements batch to be sent to Cloud DLP service in one request. */
+  public abstract Integer batchSize();
+
+  /**
+   * The transform converts the contents of input PCollection into {@link CatalogItem}s and then
+   * calls the Recommendation AI service to create the catalog item.
+   *
+   * @param input input PCollection
+   * @return PCollection after transformations
+   */
+  @Override
+  public PCollectionTuple expand(PCollection<KV<String, GenericJson>> input) {
+    return input
+        .apply("Batch Contents", ParDo.of(new BatchRequestForRecommendationAI(batchSize())))
```

Comment:
You can change this to work this way:

```suggestion
        .apply("Batch Contents", GroupIntoBatches.ofSize(batchSize())
            .withShardedKey()
            .withMaxBufferingDuration(SOMETHING))
```

where `SOMETHING` can be a default parameter, or a user-determined parameter (like batchSize). thoughts?

Reply:
Agreed. I replaced the transform with `GroupIntoBatches` and left the BufferingDuration to be set by the user (in case they want to play around with it or if there are changes to the API in the future that affect this in some way).